File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    public boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                bfs(i, computers);
                answer++;
            }
        }
        
        return answer;
    }
    
    public void bfs(int node, int[][] computers) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(node);
        visited[node] = true;
        
        while (!queue.isEmpty()) {
            int n = queue.poll();
            for (int i = 0; i < computers[n].length; i++) {
                if (computers[n][i] == 1 && !visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                }
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def bfs(computers, i, visited):
    queue = deque()
    queue.append(i)
    visited[i] = True
    
    while queue:
        node = queue.popleft()
        
        for idx in range(len(computers[node])):
            if computers[node][idx] == 1 and not visited[idx]:
                visited[idx] = True
                queue.append(idx)
    

def solution(n, computers):
    answer = 0
    visited = [False] * n
    
    for i in range(n):
        if not visited[i]:
            answer += 1
            bfs(computers, i, visited)
    
    return answer